### PR TITLE
fix(treesitter): suppress get_parser warnings via opts.error

### PR DESCRIPTION
**Pull request by ribru17**
**Problem:** #30313 introduced a change to `vim.treesitter.get_parser` that would always show a deprecation warning message without giving plugins/users a way to suppress it by updating to the new functionality.

**Solution:** Introduce `opts.error` to allow plugins/users to explicitly say that they are moving away from the deprecated functionality, thus allowing them to remove the annoying warning. See https://github.com/neovim/neovim/pull/30313#issuecomment-2349327458

### CODE REVIEW

File: runtime/lua/vim/treesitter.lua [hunk 1]
```diff
@@ -113,18 +113,26 @@ end
 ---
 --- If needed, this will create the parser.
 ---
+--- NOTE: This function will return `nil` rather than throwing an error in
+--- Nvim 0.12. Please pass `opts.error = false` to achieve this behavior
```
Comment (justinmk) [inline runtime/lua/vim/treesitter.lua:117]: a way to give ourselves a reminder about this in the version is to add a `test/functional/lua/deprecated_spec.lua` file with an `it()` test case that checks `vim.version.ge(vim.version(), '0.11')`

**Review by justinmk — APPROVED** [review]

File: runtime/lua/vim/treesitter.lua [hunk 2]
```diff
@@ -113,18 +113,26 @@ end
...
+--- now.
+---
 ---@param bufnr (integer|nil) Buffer the parser should be tied to (default: current buffer)
 ---@param lang (string|nil) Language of this parser (default: from buffer filetype)
 ---@param opts (table|nil) Options to pass to the created language tree
 ---
----@return vim.treesitter.LanguageTree object to use for parsing
+---@return vim.treesitter.LanguageTree? object to use for parsing
 function M.get_parser(bufnr, lang, opts)
   -- TODO(ribru17): Remove _get_parser and move that logic back here once the breaking function
```
Comment (echasnovski) [inline runtime/lua/vim/treesitter.lua:126]: I think there is no merit in having `_get_parser()` anymore, so maybe resolve this TODO here?

File: runtime/lua/vim/treesitter.lua [hunk 3]
```diff
@@ -438,7 +446,7 @@ end
 ---@param lang (string|nil) Language of the parser (default: from buffer filetype)
 function M.start(bufnr, lang)
   bufnr = bufnr or api.nvim_get_current_buf()
-  local parser = M._get_parser(bufnr, lang)
+  local parser = M.get_parser(bufnr, lang, { error = false })
   if not parser then
     vim.notify('No parser for the given buffer.', vim.log.levels.WARN)
```
Comment (echasnovski) [inline runtime/lua/vim/treesitter.lua:451]: I think this `vim.notify()` also suffers from the similar issue as `vim.notify_once()` in the `get_parser()`.

Prior to #30313 the call `local has_ts = pcall(vim.treesitter.start, 0)` could be used to start tree-sitter in the buffer while falling back to something else in case of `has_ts == false`. Now it does not fail silently and there is no good way of suppressing the notification. Possible solutions may be:
- Suggest to first call `vim.treesitter.get_parser` to check if there is a parser for the buffer and possibly fall back to something else. This leads to essentially calling `get_parser` twice if there *is a parser, which is not ideal.
- Show no notification if there is no parser (i.e. fail silently).

Not sure what is the best approach here. I'd probably choose a silent fail here.

Comment (clason) [reply runtime/lua/vim/treesitter.lua:451]: > I think this vim.notify() also suffers from the similar issue as vim.notify_once() in the get_parser().

Yes, but `start` is something that _users_ call (possibly interactively), so they expect feedback -- and `vim.notify` is what we have for this. Let's keep focused on the low-level `get_parser` function.

File: runtime/lua/vim/_comment.lua [hunk 4]
```diff
@@ -9,7 +9,7 @@
 local function get_commentstring(ref_position)
   local buf_cs = vim.bo.commentstring
 
-  local ts_parser = vim.treesitter._get_parser()
+  local ts_parser = vim.treesitter.get_parser(nil, nil, { error = false })
```
Comment (clason) [inline runtime/lua/vim/_comment.lua:12]: Can we pass correct arguments here? Could make it more robust...

File: runtime/lua/vim/treesitter.lua (hunk 1 above)
Comment (clason) [inline runtime/lua/vim/treesitter.lua:117]:
```suggestion
--- If no parser can be created, an error is thrown. Set `opts.error = false` to suppress this and return nil instead.
--- WARNING: This behavior will become default in Nvim 0.12 and the option will be removed.
```

File: runtime/lua/vim/treesitter.lua (hunk 2 above)
Comment (clason) [reply runtime/lua/vim/treesitter.lua:126]: Yes, that todo is obsolete now.

File: runtime/lua/vim/treesitter.lua [hunk 5]
```diff
@@ -113,18 +113,26 @@ end
...
+  opts = opts or {}
+
   local parser = M._get_parser(bufnr, lang, opts)
-  if not parser then
+  local should_error = opts.error == nil or opts.error
+  if not parser and should_error then
     vim.notify_once(
-      'WARNING: vim.treesitter.get_parser will return nil instead of raising an error in Neovim 0.12',
+      'WARNING: vim.treesitter.get_parser will return nil instead of raising an error in Neovim'
+        .. ' 0.12. Please run with "opts.error = false".',
```
Comment (clason) [inline runtime/lua/vim/treesitter.lua:135]: I thought the point was to _remove_ the warning because it can't be slurped by `pcall`?

**Review by clason — CHANGES_REQUESTED** [review]

File: runtime/lua/vim/treesitter.lua (hunk 3 above)
Comment (echasnovski) [reply runtime/lua/vim/treesitter.lua:451]: > Let's keep focused on the low-level `get_parser` function.

This was changed in #30313, so I thought it is reasonable to be discussed here.

Comment (clason) [reply runtime/lua/vim/treesitter.lua:451]: Yes, but correctly so. _That_ is where the warning _should_ be triggered (at top of the stack).

Comment (clason) [reply runtime/lua/vim/treesitter.lua:451]: But maybe I am wrong, and throwing errors and catching them (and the error message) with `pcall` is really the correct way of such "exception handling"? In this case, we should simply revert.

File: runtime/lua/vim/treesitter.lua (hunk 5 above)
Comment (ribru17) [reply runtime/lua/vim/treesitter.lua:135]: Oh, I thought the point was the make the warning remov*able* by users (by passing the argument); if the warning is removed, will it not be difficult to know (as a user) that there will be a breaking change in the future?

Comment (clason) [reply runtime/lua/vim/treesitter.lua:135]: Careful; user != caller (that would be a plugin). The point here is that we need to be careful with deprecations that concern plugins. In particular, we should absolutely minimize the chance that the warning will trigger a "press ENTER" prompt.

Personally, I would consider a soft deprecation here (doc change only).

Comment (ribru17) [reply runtime/lua/vim/treesitter.lua:135]: Ok, that makes sense

File: runtime/lua/vim/treesitter.lua (hunk 3 above)
Comment (ribru17) [reply runtime/lua/vim/treesitter.lua:451]: You both bring up great points. Is there a consensus here? I also feel like not throwing an error in `start` is breaking, as with `get_parser`

Comment (lewis6991) [reply runtime/lua/vim/treesitter.lua:451]: If we want to move away from pcall the I feel we shouldn't be calling vim.notify at all, at least for `start()`.

Comment (ribru17) [reply runtime/lua/vim/treesitter.lua:451]: I agree- I mostly meant in terms of erroring, should we still throw in `start`? I hadn't considered that this changes the behavior of `start` as well as `get_parser`

Comment (clason) [reply runtime/lua/vim/treesitter.lua:451]: There's two issues here:
1. Signaling to the calling function that a parser couldn't be found.
2. Signaling to the _user_ that a parser couldn't be found.

A change in 1. would be considered breaking (code), but a change in 2. would _not_. So no user-facing warning here should be added. If we want to prevent breaking code, we will _have to_ throw an `error`; if we don't, we simply document this and brace for impact.

Comment (ribru17) [reply runtime/lua/vim/treesitter.lua:451]: > if we don't, we simply document this and brace for impact.

Sounds good to me? :)

Comment (ribru17) [reply runtime/lua/vim/treesitter.lua:451]: Actually, I think it would be best to match the default behavior of `get_parser` (which would error until 0.12) because right now `start` returns nil even upon success. Also keeping backwards compatibility here would remove some headache

Comment (clason) [reply runtime/lua/vim/treesitter.lua]: Yes, being consistent is good. I'll think about how I want to handle this for `language.add()`, which is a lower-level API.

**Comment by ribru17** [comment]
I tried and it seems to work for me, both with this change and just with the previous `get_parser` change

**Review by clason — APPROVED** [review]
This looks fine, but is there any reason you switched from calling the private function in our code (which is exactly what it's for ;)) to calling the public function with explicit option in this PR? I don't think there was any issue with how it was (only with the user-facing `vim.notify`)?

**Comment by echasnovski** [comment]
> ... is there any reason you switched from calling the private function in our code (which is exactly what it's for ;)) to calling the public function with explicit option in this PR? ...

I think it was a reaction to [this comment](https://github.com/neovim/neovim/pull/30382#discussion_r1759985116) which *did* suggest not creating `_get_parser()` as it is mostly obsolete. I guess it is a miscommunication over the followed "Yes, that todo is obsolete now.".

File: runtime/lua/vim/treesitter.lua [hunk 6]
```diff
@@ -438,10 +426,11 @@ end
 ---@param lang (string|nil) Language of the parser (default: from buffer filetype)
 function M.start(bufnr, lang)
   bufnr = bufnr or api.nvim_get_current_buf()
-  local parser = M._get_parser(bufnr, lang)
+  local parser = M.get_parser(bufnr, lang, { error = false })
```
Comment (clason) [inline runtime/lua/vim/treesitter.lua:429]: use `assert` here?

**Comment by clason** [comment]
> > ... is there any reason you switched from calling the private function in our code (which is exactly what it's for ;)) to calling the public function with explicit option in this PR? ...
> 
> I think it was a reaction to [this comment](https://github.com/neovim/neovim/pull/30382#discussion_r1759985116) which _did_ suggest not creating `_get_parser()` as it is mostly obsolete. I guess it is a miscommunication over the followed "Yes, that todo is obsolete now.".

Ah, ok, I missed that `_get_parser` is gone completely (change view was confusing...)

That's fine, but since we don't have a standard deprecation warning, we should probably still have a `todo` about removing the option -- or is the documentation warning enough?

File: runtime/lua/vim/treesitter.lua (hunk 6 above)
Comment (echasnovski) [reply runtime/lua/vim/treesitter.lua:429]: Later `if not parser then` gives a more informative error message.

Comment (clason) [reply runtime/lua/vim/treesitter.lua:429]: No, I mean use `assert` _here_ with the more informative error message. It's equivalent, of course; it's just a question of which pattern we want to promote `if not foo then error('bar')` vs. `assert(foo, 'bar')`).

File: runtime/doc/treesitter.txt [hunk 7]
```diff
@@ -818,6 +818,10 @@ get_parser({bufnr}, {lang}, {opts})              *vim.treesitter.get_parser()*
 
     If needed, this will create the parser.
 
+    If no parser can be created, an error is thrown. Set `opts.error = false`
+    to suppress this and return nil instead. WARNING: This behavior will
+    become default in Nvim 0.12 and the option will be removed.
```
Comment (gpanders) [inline runtime/doc/treesitter.txt:823]: The option becoming default and the option being removed should happen in separate releases. Might want to reword here to clarify that. “This behavior will become default in Nvim 0.12 and removed in a future release”

Comment (clason) [reply runtime/doc/treesitter.txt:823]: No, the proposal is in fact for an expedited deprecation (flying under the "experimental" flag). If there are _specific_ objections to that (and the related API change; this is a bigger project), I would like to hear them. (This change is _much_ easier to adapt to than the query iteration one.)

Comment (clason) [reply runtime/doc/treesitter.txt:823]: To be more precise: What I'm looking for here is scenarios where the option removal will be breaking whereas the default switch would not be. (Note that you're opting in to the new behavior, not the old one, from the start -- unlike the query change.)

**Comment by ribru17** [comment]
> That's fine, but since we don't have a standard deprecation warning, we should probably still have a `todo` about removing the option -- or is the documentation warning enough?

I added a test indicating the option should be removed in 0.12, maybe that + the documentation warning is enough?

File: runtime/lua/vim/treesitter.lua [hunk 8]
```diff
@@ -438,11 +426,10 @@ end
...
 function M.start(bufnr, lang)
   bufnr = bufnr or api.nvim_get_current_buf()
-  local parser = M._get_parser(bufnr, lang)
-  if not parser then
-    vim.notify('No parser for the given buffer.', vim.log.levels.WARN)
-    return
-  end
+  local parser = assert(
+    M.get_parser(bufnr, lang, { error = false }),
+    string.format('Highlighting could not be started for buffer: %s. No parser found.', bufnr)
```
Comment (clason) [inline runtime/lua/vim/treesitter.lua:431]: Sorry to "one more thing" you, but I think the error message should contain as much context as possible -- in this case also the language name. (This may be called from "distros", so the context may not be always clear to the user.)

(This is related to a regression on treesitter highlight error messages, which used to contain the language and query name.)

File: runtime/lua/vim/treesitter/dev.lua [hunk 9]
```diff
@@ -76,7 +76,7 @@ end
 ---
 ---@package
 function TSTreeView:new(bufnr, lang)
-  local parser = vim.treesitter._get_parser(bufnr or 0, lang)
+  local parser = vim.treesitter.get_parser(bufnr or 0, lang, { error = false })
   if not parser then
     return nil, 'No parser available for the given buffer.'
   end
```
Comment (clason) [inline runtime/lua/vim/treesitter/dev.lua:82]: can this be an assert (with context) as well?

File: runtime/lua/vim/treesitter/dev.lua [hunk 10]
```diff
@@ -595,7 +595,7 @@ function M.edit_query(lang)
...
   vim.cmd(cmd)
 
-  local parser = vim.treesitter._get_parser(buf, lang)
+  local parser = vim.treesitter.get_parser(buf, lang, { error = false })
   if not parser then
     return nil, 'No parser available for the given buffer'
   end
```
Comment (clason) [inline runtime/lua/vim/treesitter/dev.lua:601]: same?

Comment (clason) [reply runtime/lua/vim/treesitter/dev.lua:601]: (sorry for the iterations; I want to get this right, since it'll establish a pattern for use elsewhere, e.g., my `add()` PR.)

File: runtime/lua/vim/treesitter/dev.lua (hunk 9 above)
Comment (gpanders) [reply runtime/lua/vim/treesitter/dev.lua:82]: I think we should show an error in 
`:InspectTree`, but not here in the API.

Comment (clason) [reply runtime/lua/vim/treesitter/dev.lua:82]: And I am actually not at all fond of propagating error messages as strings. Either
1. `assert` them,
2. notify them, or 
3. log them,

but keep them where they happen.

Comment (gpanders) [reply runtime/lua/vim/treesitter/dev.lua:82]: > And I am actually not at all fond of propagating error messages as strings.

It's a standard Lua pattern. The error should be asserted/notified at the application level (`:InspectTree` in this case, or other plugin code), not the "library"/API level.

Comment (clason) [reply runtime/lua/vim/treesitter/dev.lua:82]: Yes, and pcall is the standard approach for that pattern (throw/catch), isn't it?

Comment (clason) [reply runtime/lua/vim/treesitter/dev.lua:82]: Sorry, I'm not making my point very well; let me try again. It's not about the error itself, it's about the message -- in particular, where it is constructed. I believe that this should be as close as possible to the point where the actual error is raised.

In this case, you're asking a function "do this thing", to which it replies "I cannot do this thing you ask". Now, we have just established a convention for that: simply return `nil`, which tells the calling function all it needs to know. Why should we return an additional string that says exactly the same thing, just in (more) words? What is that message telling the calling function that it cannot infer from the `nil` return, and what does it do with this information? Why can't `:InspectTree` not simply do `assert(parser, 'No parser available for the buffer')`?

(Or put differently: Hard-coded error strings without dynamic content are sus.)

File: runtime/lua/vim/treesitter/dev.lua (hunk 10 above)
Comment (clason) [reply runtime/lua/vim/treesitter/dev.lua:601]: Concretely, my proposal is 'do not return an error message here, `nil` already means "No parser available", just return'.

Comment (gpanders) [reply runtime/lua/vim/treesitter/dev.lua:601]: How else will the caller know why the function failed? Maybe there is only a single error path now, but there could be more in the future, and in that case we would want to explain what the error was. 

Returning an error message also lets callers simply use `local blah = assert(vim.treesitter.dev.show_tree())` which will conveniently show the error message “for free” (this is the same pattern Lua itself uses in e.g. the `io` library. I don’t see any reason to deviate here).

Comment (clason) [reply runtime/lua/vim/treesitter/dev.lua:601]: Then that error message should provide more value, if it's the (only) one that is surfaced to the user. (At the very least, _which_ parser? _which_ buffer?)

Comment (clason) [reply runtime/lua/vim/treesitter/dev.lua:601]: Also, I'm a bit confused, since we're commenting here on `:EditQuery`, which does not have a "caller" (in the sense of a calling function -- unless we are talking about the `nvim_create_usercommand`, which then needs to `assert` the call).

Sorry for being so ornery, but I'm genuinely interested in having a discussion about this so I can understand the _full_ pattern across our API (at least treesitter, possibly beyond). Ideally, we'll get something we can actually write in `develop.txt` and I can follow in future PRs instead of winging it based on contributor style as we (certainly I) have so far.

(Part of the concern is that adding the error message to the return value locks the return part of the API. If we want to promote this pattern, we'll have to standardize on `retval, err`, with `retval` being a table for forward compatibility. Of course, that's very restrictive...)

Comment (gpanders) [reply runtime/lua/vim/treesitter/dev.lua:601]: >Also, I'm a bit confused, since we're commenting here on :EditQuery, which does not have a "caller" (in the sense of a calling function -- unless we are talking about the nvim_create_usercommand, which then needs to assert the call).

No, we're talking about `vim.treesitter.dev.edit_query()`, which does have a caller.

>If we want to promote this pattern, we'll have to standardize on retval, err, with retval being a table for forward compatibility. 

This is a good point, and I hadn't appreciated until this comment that the API in question (`edit_query`) does not return a value.

IMO for functions which are only called for their side effects (like this one), calling `assert` or `error` directly is appropriate (I realize that takes this entire discussion in a different direction, I apologize for not looking closer at this earlier).

The pattern of returning `retval, err` is the right one to use for functions which return a value (like `TSTreeView::new`).

Comment (clason) [reply runtime/lua/vim/treesitter/dev.lua:601]: Thank you, that helps in wrapping my mind around this. It's clear that "one size fits all" fits none, so my goal is to understand (with the help of these examples) which pattern is appropriate in which situation; and the difference between "generators" (that return a value) and "actors" (that don't, and are called purely for their side effects) is one I tried making earlier.

What muddies the waters here unfortunately is that some of the functions change their type in these PRs...

Comment (mfussenegger) [reply runtime/lua/vim/treesitter/dev.lua:601]: >  but I'm genuinely interested in having a discussion about this so I can understand the full pattern across our API (at least treesitter, possibly beyond). Ideally, we'll get something we can actually write in develop.txt and I can follow in future PRs instead of winging it based on contributor style as we (certainly I) have so far.

> (Part of the concern is that adding the error message to the return value locks the return part of the API. If we want to promote this pattern, we'll have to standardize on retval, err, with retval being a table for forward compatibility. Of course, that's very restrictive...)


I think part of the challenge here is that:

- there's no clear established pattern (yet)
- it's not one size fit all
- different contributors have slightly different preferences

From what I can tell we have at least:

1. `error("no luck")`
2. `retval, errmsg`
3. optional result, no errormsg
4. `on_error` parameter (async, visitors - where `n` errors can happen)
5. `vim.notify` (sometimes with optional `opts.silent` (async, visitors ^))


My personal reasoning for when to use which (so far):


1) is well suited for cases where the error is very exceptional, in a lower level and where the consumers are unlikely to handle it in a meaningful way. Advantage over 2) is that propagation happens for free and it's harder to accidentally swallow errors. (E.g. using `uv_handle/pipe:write()` without checking return values is common)

2) is well suited for cases where the error is more likely or where having a fallback is more realistic.

3) Kinda same as 2. where there is only a single case of "doesn't exist" (e.g. cache lookup, dict lookup)

4) lower level, either async methods or where the error can happen multiple times

5) Higher level. End-user invokes these directly

Comment (clason) [reply runtime/lua/vim/treesitter/dev.lua:601]: Thank you, that is exactly the sort of guidance I was hoping for (and puts a proper face on my vague and confusing thoughts).

Comment (mfussenegger) [reply runtime/lua/vim/treesitter/dev.lua:601]: > what would be guidelines for using pcall / assert? (Eating vs propagating vs. throwing errors).

I'd say:

`pcall` only if you either:

- rethrow and enrich the error with additional context
- can provide a fallback, or in case of a higher level function want to avoid the user getting a error with stacktrace but only a warning

`assert` is either:

- defensive programming: To make an assumption explicit that should always be true due to contracts.
- alternative to `error` ⬇️

In regards to `assert(truthy, msg)` vs. `if condition then error(msg) end`:
Use the latter if building the `msg` would be costly (and do it only within the if block. Otherwise `assert` is fine.


(Disclaimer: This is just how I've been using these things, not sure if there's consensus on these)

Comment (clason) [reply runtime/lua/vim/treesitter/dev.lua:601]: Ah, that's a good point: `assert` needs a constructed message (which I believe should be the default ;))

Comment (justinmk) [reply runtime/lua/vim/treesitter/dev.lua:601]: Incorporated @mfussenegger 's outline into https://github.com/neovim/neovim/pull/30240

**Comment by clason** [comment]
@ribru17 sorry for holding this up -- could you change the errors/asserts to align with @mfussenegger's suggestions (in particular, include the context (language) in the messages?) Then I think we can merge.

(If you prefer not to do it, let me know and I'll add a todo instead and try to address it in my follow-up PR.)

**Comment by justinmk** [comment]
> you change the errors/asserts to align with mfussenegger's suggestions

After https://github.com/neovim/neovim/pull/30240 these will be documented in `:help dev-error-patterns`.

**Comment by ribru17** [comment]
@clason No worries, sorry for the delay I'll fix this up today once I am at my computer

**Comment by ribru17** [comment]
There was a lot of discussion here and I want to make sure I get this right- the consensus is to use `assert` in `edit_query()` and return `nil, err_msg` (with a more descriptive, dynamic error message) in `TSTreeView:new`?

File: runtime/lua/vim/treesitter.lua (hunk 8 above)
Comment (ribru17) [reply runtime/lua/vim/treesitter.lua]: Let me know if this change is not acceptable; I think the behavior is slightly less intuitive now because often `start()` is passed `nil` and the language is determined in the `get_parser` side, but now we shoot out the error message in `start()` (before that `lang` calculation) so the error message may be less helpful in a lot of cases (i.e. cases where you are *not* passing the language with something like `vim.treesitter.start(0, 'javascript')`

**Comment by clason** [comment]
> There was a lot of discussion here and I want to make sure I get this right- the consensus is to use `assert` in `edit_query()` and 

No, I think the suggestion was to return `nil, err` in `edit_query()` and `assert` in `query.edit()` (I missed that little bit of indirection, hence the confusion).

> return `nil, err_msg` (with a more descriptive, dynamic error message) in `TSTreeView:new`?

Correct, and `assert` in `inspect_tree()` (which one?)

@gpanders did I get that right?

**Comment by ribru17** [comment]
> (And make sure to build the error message conditionally on failure.)

Gotcha, so I'm guessing this error should be of the form `if condition then error(msg) end`

**Comment by clason** [comment]
or `if condition then return nil, string.format(...) end` (rather than having the format outside the condition)

**Comment by ribru17** [comment]
Alright, I think this is good to go- I made the last commit a fixup so it would be easier to see what changed. Looks like `inspect_tree` already handled assertion logic (around `TSTreeView`) so I was able to leave that unchanged

I think in the future it would be good to have `edit_query` return `true` on success so that way there is a reason to return `nil` in the first argument of the failure case (right now it's weird because it will return `nil` anyway)

File: runtime/lua/vim/treesitter.lua [hunk 11]
```diff
@@ -438,11 +426,14 @@ end
...
 function M.start(bufnr, lang)
   bufnr = bufnr or api.nvim_get_current_buf()
-  local parser = M._get_parser(bufnr, lang)
-  if not parser then
-    vim.notify('No parser for the given buffer.', vim.log.levels.WARN)
-    return
-  end
+  local parser = assert(
+    M.get_parser(bufnr, lang, { error = false }),
+    string.format(
```
Comment (clason) [inline runtime/lua/vim/treesitter.lua:431]: I think that is a LuaJIT-only signature. (And constructing the error message inside an `assert` is exactly what we shouldn't do -- we should just use the message that `get_parser` constructs in the fail branch so we can just `assert` it.)

File: runtime/lua/vim/treesitter.lua [hunk 12]
```diff
@@ -94,11 +99,17 @@ function M._get_parser(bufnr, lang, opts)
 
   if not valid_lang(lang) then
     if not parsers[bufnr] then
+      if should_error then
+        error('Parser not found.')
+      end
       return nil
```
Comment (clason) [inline runtime/lua/vim/treesitter.lua:105]: ´return nil, errmsg` and `error(errmsg)`, same below.

File: runtime/lua/vim/treesitter/dev.lua [hunk 13]
```diff
@@ -595,9 +600,14 @@ function M.edit_query(lang)
...
   vim.cmd(cmd)
 
-  local parser = vim.treesitter._get_parser(buf, lang)
+  local parser = vim.treesitter.get_parser(buf, lang, { error = false })
   if not parser then
-    return nil, 'No parser available for the given buffer'
+    return nil,
+      string.format(
+        'Failed to show query editor for buffer: %s. No parser for lang: %s.',
```
Comment (clason) [inline runtime/lua/vim/treesitter/dev.lua:607]:
```suggestion
        'Failed to show query editor for buffer %s: no parser for lang %s.',
```

Nit, while something still has to be fixed anyway. (And similarly in other places; colons are not needed here.)

File: runtime/lua/vim/treesitter/_fold.lua [hunk 14]
```diff
@@ -114,7 +114,7 @@ local function compute_folds_levels(bufnr, info, srow, erow, parse_injections)
   srow = srow or 0
   erow = erow or api.nvim_buf_line_count(bufnr)
 
-  local parser = assert(ts._get_parser(bufnr))
+  local parser = assert(ts.get_parser(bufnr, nil, { error = false }))
```
Comment (clason) [inline runtime/lua/vim/treesitter/_fold.lua:117]: Can this assert even be triggered, given that  `foldexpr` already returns early on the same check?

(That'd be fine and a good use of `assert` without error message; just wondering.)

File: runtime/lua/vim/treesitter.lua (hunk 11 above)
Comment (clason) [reply runtime/lua/vim/treesitter.lua:431]: Oh, wait, `bufnr` and especially `lang` are only determined in `get_parser`. Then we _have_ to use the `nil, errmsg` pattern and just `assert(get_parser(...))` here!

File: runtime/lua/vim/treesitter/_fold.lua (hunk 14 above)
Comment (ribru17) [reply runtime/lua/vim/treesitter/_fold.lua:117]: I think it maybe could, due to being called in the `autocmd` at the bottom. Not certain but maybe best to add an error message just in case

File: runtime/lua/vim/treesitter.lua [hunk 15]
```diff
@@ -94,12 +100,21 @@ function M._get_parser(bufnr, lang, opts)
...
+        error(err_msg)
+      end
+      return nil, err_msg
     end
   elseif parsers[bufnr] == nil or parsers[bufnr]:lang() ~= lang then
     local parser = vim.F.npcall(M._create_parser, bufnr, lang, opts)
     if not parser then
-      return nil
+      local err_msg =
+        string.format('Parser could not be created for buffer %s, lang "%s"', bufnr, lang)
```
Comment (clason) [inline runtime/lua/vim/treesitter.lua:113]:
```suggestion
        string.format('Parser could not be created for buffer %s and language "%s"', bufnr, lang)
```

File: runtime/lua/vim/treesitter/dev.lua [hunk 16]
```diff
@@ -595,9 +600,10 @@ function M.edit_query(lang)
   end
   vim.cmd(cmd)
 
-  local parser = vim.treesitter._get_parser(buf, lang)
+  local parser = vim.treesitter.get_parser(buf, lang, { error = false })
```
Comment (clason) [inline runtime/lua/vim/treesitter/dev.lua:603]: could `assert` here a well

File: runtime/lua/vim/treesitter/query.lua [hunk 17]
```diff
@@ -1028,7 +1028,10 @@ end
 ---
 --- @param lang? string language to open the query editor for. If omitted, inferred from the current buffer's filetype.
 function M.edit(lang)
-  vim.treesitter.dev.edit_query(lang)
+  local _, err_msg = vim.treesitter.dev.edit_query(lang)
```
Comment (clason) [inline runtime/lua/vim/treesitter/query.lua:1031]: assert?

File: runtime/lua/vim/treesitter.lua [hunk 18]
```diff
@@ -94,12 +100,21 @@ function M._get_parser(bufnr, lang, opts)
 
   if not valid_lang(lang) then
     if not parsers[bufnr] then
-      return nil
+      local err_msg = string.format('Parser not found for buffer %s: invalid language', bufnr)
```
[the remark this answers is not in the document]
Comment (ribru17) [reply runtime/lua/vim/treesitter.lua:103]: I think maybe it should be changed to `language could not be determined`? It isn't clear from the code (and I didn't make it clear enough here) that `lang` will always be `nil` right here

File: runtime/lua/vim/treesitter/query.lua (hunk 17 above)
Comment (ribru17) [reply runtime/lua/vim/treesitter/query.lua:1031]: I thought about that but I think it cannot be done here since `edit_query` returns `nil` even upon success (meaning the assertion would always fail). That's why I think it should return true on success but maybe could be done in a small follow-up PR? To lubricate the merging of this one

File: runtime/lua/vim/treesitter/dev.lua (hunk 16 above)
Comment (ribru17) [reply runtime/lua/vim/treesitter/dev.lua:603]: I thought `edit_query` was meant to return `nil, err_msg` while `query.edit` should do the actual assertion?

File: runtime/lua/vim/treesitter.lua (hunk 18 above)
[the remark this answers is not in the document]
Comment (clason) [reply runtime/lua/vim/treesitter.lua]: Yes, thank you, I missed that!

File: runtime/lua/vim/treesitter/dev.lua (hunk 16 above)
Comment (clason) [reply runtime/lua/vim/treesitter/dev.lua:603]: Right, sorry, I keep getting these two mixed up. I see now that you return here rather than `error`, so this is fine of course.

File: runtime/lua/vim/treesitter/query.lua (hunk 17 above)
Comment (clason) [reply runtime/lua/vim/treesitter/query.lua:1031]: Please add a todo then. (And let's follow up quickly ;))

**Review by justinmk — APPROVED** [review]
